android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/Utils.kt

```diff
@@ -11,49 +11,51 @@ private const val SEARCH_TYPE_SAP_FOLLOW_ON = "sap-follow-on"
 private const val SEARCH_TYPE_SAP = "sap"
 private const val SEARCH_TYPE_ORGANIC = "organic"
 private const val CHANNEL_KEY = "channel"
-private val validCodeSet = setOf(
-    "MOZ2", "MOZ4", "MOZ5", "MOZA", "MOZB", "MOZD", "MOZE", "MOZI", "MOZM", "MOZO", "MOZT",
-    "MOZW", "MOZSL01", "MOZSL02", "MOZSL03", "monline_dg", "monline_3_dg", "monline_4_dg",
-    "monline_7_dg", "firefox-a", "firefox-b", "firefox-b-1", "firefox-b-ab", "firefox-b-1-ab",
-    "firefox-b-d", "firefox-b-1-d", "firefox-b-e", "firefox-b-1-e", "firefox-b-m",
-    "firefox-b-1-m", "firefox-b-o", "firefox-b-1-o", "firefox-b-lm", "firefox-b-1-lm",
-    "firefox-b-lg", "firefox-b-huawei-h1611", "firefox-b-is-oem1", "firefox-b-oem1",
-    "firefox-b-oem2", "firefox-b-tinno", "firefox-b-pn-wt", "firefox-b-pn-wt-us", "ubuntu",
-    "ffab", "ffcm", "ffhp", "ffip", "ffit", "ffnt", "ffocus", "ffsb", "fpas",
-    "newext", "1000969a", null,
-)
 private val validChannelSet = setOf("ts")
 
 /**
  * Get a String in a specific format allowing to identify how an ads/search provider was used.
  *
  * @see [TrackKeyInfo.createTrackKey]
  */
+@Suppress("NestedBlockDepth", "ComplexMethod")
 internal fun getTrackKey(
     provider: SearchProviderModel,
     uri: Uri,
     cookies: List<JSONObject>,
 ): String {
+    var type = SEARCH_TYPE_ORGANIC
     val paramSet = uri.queryParameterNames
     var code: String? = null
 
-    if (provider.codeParam.isNotEmpty()) {
-        code = uri.getQueryParameter(provider.codeParam)
+    if (provider.codeParamName.isNotEmpty()) {
+        code = uri.getQueryParameter(provider.codeParamName)
         if (code.isNullOrEmpty() &&
-            provider.name == "baidu" &&
+            provider.telemetryId == "baidu" &&
             uri.toString().contains("from=")
         ) {
             code = uri.toString().substringAfter("from=", "")
                 .substringBefore("/", "")
         }
+        if (!code.isNullOrEmpty()) {
+            if (provider.taggedCodes.contains(code)) {
+                type = SEARCH_TYPE_SAP
+                if (!provider.followOnParamNames.isNullOrEmpty() &&
+                    provider.followOnParamNames.any { prefix ->
+                        paramSet.contains(prefix)
+                    }
+                ) {
+                    type = SEARCH_TYPE_SAP_FOLLOW_ON
+                }
+            } else if (provider.organicCodes?.contains(code) == true) {
+                type = SEARCH_TYPE_ORGANIC
+            } else if (provider.expectedOrganicCodes?.contains(code) == true) {
+                code = "none"
+            } else {
+                code = "other"
+            }
```

Comment:
I think it would be useful in terms of reader clarity to have these pulled into a separate function that uses a `when` statement instead. Having a function name to describe this block might go a long way 🙂 Hopefully that will help with the lint checks that you've suppressed above as well

Reply:
+1 We should absolutely try to avoid `NestedBlockDepth` and `ComplexMethod` especially considering this is our search code and we would be best served if this is easy to test and easily understood. 

I would suggest trying to figure out if you can remove some of the if statement nesting by doing an early return. Here's a sample change to remove the top level if statement nesting

```
var code: String? = "none"

if (provider.codeParamName.isEmpty() {
   return TrackKeyInfo(provider.telemetryId, type, code).createTrackKey()
}

code = uri.getQueryParameter(provider.codeParamName)
if (code.isNullOrEmpty() &&
    provider.telemetryId == "baidu" &&
    uri.toString().contains("from=")
) {
  ... 
}
```
I believe you could declare `code` to default to `none` from the get go so you don't need the extra if statement at the end.